### PR TITLE
#440: Re-enable HTML report after introduction of test plans.

### LINKED ISSUE #440: restore HTML report and Graphs
PR https://github.com/tag1consulting/goose/pull/422 temporarily disabled `--report-file` which needs to be updated to properly support the new Test Plan functionality. The HTML report should be able to generally duplicate `fmt_overview`. The graphs will need to be reworked as `GoosePhase::Starting`, `::Running` and `::Stopping` have been replaced with `::Increase`, `::Maintain` and `::Decrease`, each of which can happen multiple times during a given test plan.

### PR DESCRIPTION
**Pull request by slashrsm**
Closes #440. 

- Re-enables report file
- Instead of starting, started, stopping, stopped summary displays a table of plan steps overview
- Removes "Starting" and "Stopping" section from graphs
- Add increasing (green) and decreasing (red) sections to the graph

Screenshot 2022-04-21 at 12-24-03 Goose Attack Report

### CODE REVIEW

File: src/metrics.rs [hunk 1]
```diff
@@ -2607,50 +2602,61 @@ impl GooseAttack {
...
-                report_range.push_str(&format!(
-                    "<p>Running: <span>{} - {} (Duration: {:02}:{:02}:{:02})</span></p>",
-                    started.format("%Y-%m-%d %H:%M:%S"),
-                    stopping.format("%Y-%m-%d %H:%M:%S"),
-                    running_hours,
-                    running_minutes,
-                    running_seconds,
-                ));
+            let mut steps_overview = String::new();
+            for step in self.metrics.history.windows(2) {
```
Comment (slashrsm) [inline src/metrics.rs:2606]: This loop is basically a copy-paste of the same thing in console printout part. Maybe we should move it to a helper and try to share it?

File: src/lib.rs [hunk 2]
```diff
@@ -1733,15 +1735,15 @@ impl GooseAttack {
...
                     self.sync_metrics(&mut goose_attack_run_state, true).await?;
                     // Record last users for users per second graph in HTML report.
-                    //self.graph_data.record_users_per_second(self.metrics.users, Utc::now());
+                    self.graph_data
+                        .record_users_per_second(self.metrics.users, Utc::now());
                     // The load test is fully stopped at this point.
                     self.metrics
                         .history
                         .push(TestPlanHistory::step(TestPlanStepAction::Finished, 0));
                     //self.graph_data.set_stopped(Utc::now());
```
Comment (jeremyandrews) [inline src/lib.rs:1744]: This is no longer needed, delete.

File: src/graph.rs [hunk 3]
```diff
@@ -16,14 +17,7 @@ use std::marker::PhantomData;
 /// Used to collect graph data during a load test.
 pub(crate) struct GraphData {
     /// Tracks when the load test first started with an optional system timestamp.
-    starting: Option<DateTime<Utc>>,
-    /// Tracks when all [`GooseUser`](../goose/struct.GooseUser.html) threads fully
-    /// started with an optional system timestamp.
     started: Option<DateTime<Utc>>,
```
Comment (jeremyandrews) [inline src/graph.rs:20]: Why is this needed? Is this missing in `metrics.history`?

File: src/graph.rs [hunk 4]
```diff
@@ -53,26 +44,11 @@ impl GraphData {
         }
     }
 
-    /// Sets starting time.
-    pub(crate) fn set_starting(&mut self, starting: DateTime<Utc>) {
-        self.starting = Some(starting);
-    }
-
     /// Sets started time.
     pub(crate) fn set_started(&mut self, started: DateTime<Utc>) {
```
Comment (jeremyandrews) [inline src/graph.rs:48]: I think this is unnecessary now?

**Review by jeremyandrews — CHANGES_REQUESTED** [review]
I don't think we need to track `started` a second time, but can instead get it from `metrics.history`?

File: src/lib.rs (hunk 2 above)
Comment (jeremyandrews) [inline src/lib.rs:1744]: `//self.graph_data.set_stopped(Utc::now());`

It seems this isn't used anymore, so can be removed?

Comment (slashrsm) [reply src/lib.rs]: Done.

File: src/graph.rs (hunk 3 above)
Comment (slashrsm) [reply src/graph.rs]: Removed

File: src/graph.rs (hunk 4 above)
Comment (slashrsm) [reply src/graph.rs]: Indeed.

File: src/lib.rs [hunk 5]
```diff
@@ -78,12 +78,13 @@ use tokio::fs::File;
...
 use crate::controller::{GooseControllerProtocol, GooseControllerRequest};
 use crate::goose::{GaggleUser, GooseTask, GooseTaskSet, GooseUser, GooseUserCommand};
-//use crate::graph::GraphData;
+use crate::graph::GraphData;
 use crate::logger::{GooseLoggerJoinHandle, GooseLoggerTx};
 use crate::metrics::{GooseMetric, GooseMetrics};
 use crate::test_plan::{TestPlan, TestPlanHistory, TestPlanStepAction};
 #[cfg(feature = "gaggle")]
 use crate::worker::{register_shutdown_pipe_handler, GaggleMetrics};
+use chrono::prelude::*;
```
[the remark this answers is not in the document]
Comment (slashrsm) [reply src/lib.rs]: This is gone now.

File: src/metrics.rs [hunk 6]
```diff
@@ -2604,61 +2599,73 @@ impl GooseAttack {
...
 
-    /* @FIXME:
     // Write an HTML-formatted report, if enabled.
     pub(crate) async fn write_html_report(
         &mut self,
         goose_attack_run_state: &mut GooseAttackRunState,
     ) -> Result<(), GooseError> {
         // Only write the report if enabled.
         if let Some(report_file) = goose_attack_run_state.report_file.as_mut() {
+            let test_start_time = self.metrics.history.first().unwrap().timestamp;
```
Comment (jeremyandrews) [inline src/metrics.rs:2609]: I try to use `.expect()` instead of `.unwrap()` to more explicitly document our assumptions -- but not a blocker as it will panic either way.

**Review by jeremyandrews — APPROVED** [review]
Great work!
